### PR TITLE
#5206 When using ReactorBuilder, twisted.internet.reactor points to the temporary reactor

### LINKED ISSUE #5206: ReactorBuilder test infrastructure breaks code that uses the global reactor
|[itamarst's avatar](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5206|
|Type|defect|
|Created|2011-07-27 11:52:15Z|

Code that uses the global reactor, e.g. `twisted.internet.task.cooperate`, will not work correctly with tests that use the ReactorBuilder infrastructure. While such code should be updated to have an explicit reactor passed in, doing so can be non-trivial (e.g. `cooperate` needs a registry, or it's a 3rd party library that cannot be modified).

Providing some workaround to this would both be helpful for testing, but also (depending on the implementation) could later be expanded to help catch code that relies on the global reactor instance.

<details><summary>Searchable metadata</summary>

```
trac-id__5206 5206
type__defect defect
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__core core
keywords__ 
time__1311767535000000 1311767535000000
changetime__1321499609000000 1321499609000000
version__None None
owner__itamar itamar

```
</details>

### PR DESCRIPTION
**Pull request by itamarst**
## Scope and purpose

Fixes #5206

Working on #11989, I end up fighting the fact that the TLS classes don't have a reference to the reactor, so I need to rely on the global reactor, which then hits the limitation of ReactorBuilder that it doesn't affect the global reactor.

This PR makes ReactorBuilder affect the global reactor. It's horrible, I admit, but the alternative is lots of patching.

## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/core/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/core/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

### CODE REVIEW

File: src/twisted/python/test/strategies.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ def systemdDescriptorNames() -> SearchStrategy[str]:
             min_codepoint=0,
             max_codepoint=127,
             # This one excludes control characters.
-            blacklist_categories=("Cc",),
+            blacklist_categories=["Cc"],
```
Comment (itamarst) [inline src/twisted/python/test/strategies.py]: This is unrelated fix that makes `mypy` happy with newer Hypothesis.

File: src/twisted/internet/test/test_tls.py [hunk 2]
```diff
@@ -390,16 +389,6 @@ class AbortSSLConnectionTests(
     requiredInterfaces = (IReactorSSL,)
     endpoints = SSLCreator()
 
-    def buildReactor(self):
```
Comment (itamarst) [inline src/twisted/internet/test/test_tls.py:393]: No longer necessary since we're now making the global reactor match the temporary one.

**Comment by glyph** [comment]
Could the TLS protocol classes be adapted, instead, to use something like `.transport.reactor`, to avoid doing this?  ReactorBuilder not supporting patching the global reactor is one of the few subtle pressures we've got to encourage maintenance to slowly remove global reactor references.

**Comment by itamarst** [comment]
Is `.transport.reactor` guaranteed to work on all transports and/or all reactor implementations? It's not part of the `ITransport` interface, so there's no reason to think it'll be there.

**Comment by glyph** [comment]
> Is `.transport.reactor` guaranteed to work on all transports and/or all reactor implementations? It's not part of the `ITransport` interface, so there's no reason to think it'll be there.

It absolutely isn't! But is `__dict__` assignment guaranteed to work across all reactors? ;-)

**Comment by itamarst** [comment]
The difference is that `__dict__` assignment only happens in the test suite, but `.transport.reactor` would have to happen in non-test code.

**Comment by itamarst** [comment]
How about I do the other PR without relying on this, and then it'll be clearer why this is useful.

File: src/twisted/internet/test/reactormixins.py [hunk 3]
```diff
@@ -247,6 +247,13 @@ def unbuildReactor(self, reactor):
         for c in calls:
             c.cancel()
 
+        if global_reactor_was_munged:
```
Comment (adiroiban) [inline src/twisted/internet/test/reactormixins.py]: Just a minor comment.
Maybe it would be better to have this as a separate function called `unbuildGlobalReactor`

I think that having a call like

```
self.addCleanup(self.unbuildReactor, reactor)
self.addCleanup(self.unbuildGlobalReactor, reactor)
```

is easier to read than the "magic"

```
self.addCleanup(self.unbuildReactor, reactor)
```

----

Basically, this is the reverse, with `global_reactor_was_munged=False` by default

File: src/twisted/internet/test/test_gireactor.py [hunk 4]
```diff
@@ -104,7 +104,7 @@ def test_gApplicationActivate(self) -> None:
         L{Gio.Application} instances can be registered with a gireactor.
         """
         reactor = gireactor.GIReactor(useGtk=False)
-        self.addCleanup(self.unbuildReactor, reactor)
+        self.addCleanup(self.unbuildReactor, reactor, False)
```
Comment (adiroiban) [inline src/twisted/internet/test/test_gireactor.py]: I can see that the test is executed on linux and the coverage was published for that run... maybe a codecov.io bug

image

**Review by adiroiban — APPROVED** [review]
Thanks for the cleanup.

Only minor comment regarding using the flag argument vs a separate function. 

As this is just test code, I think we can merge it.

If anyone find something wrong with this code, we can change it later.

File: src/twisted/internet/test/reactormixins.py [hunk 5]
```diff
@@ -216,7 +216,7 @@ def tearDown(self):
                         % (process.reapProcessHandlers,)
                     )
 
-    def unbuildReactor(self, reactor):
+    def unbuildReactor(self, reactor, global_reactor_was_munged=True):
```
Comment (glyph) [inline src/twisted/internet/test/reactormixins.py]:
```suggestion
    def unbuildReactor(self, reactor, globalReactorWasMunged=True):
```

File: src/twisted/internet/test/reactormixins.py (hunk 3 above)
Comment (glyph) [inline src/twisted/internet/test/reactormixins.py]:
```suggestion
        if globalReactorWasMunged:
```

**Review by glyph — COMMENTED** [review]
(Just a couple of coding standard inline suggestions for your convenience. I don't disagree with Adi's feedback, if you want to go that route.)

File: src/twisted/internet/test/reactormixins.py (hunk 3 above)
Comment (itamarst) [reply src/twisted/internet/test/reactormixins.py]: Hm. It is a little confusing as is. Perhaps a better solution is making `test_gireactor.py` work better, since that's the only use case for the flag.

Comment (itamarst) [reply src/twisted/internet/test/reactormixins.py]: I ended making `unbuildReactor` private.

File: src/twisted/internet/test/test_gireactor.py [hunk 6]
```diff
@@ -103,8 +103,8 @@ def test_gApplicationActivate(self) -> None:
         """
         L{Gio.Application} instances can be registered with a gireactor.
         """
-        reactor = gireactor.GIReactor(useGtk=False)
-        self.addCleanup(self.unbuildReactor, reactor)
+        self.reactorFactory = lambda: gireactor.GIReactor(useGtk=False)
+        reactor = self.buildReactor()
```
Comment (adiroiban) [inline src/twisted/internet/test/test_gireactor.py:107]: The refactoring looks very good.

Just a minor suggestion:
```suggestion
        reactor = self.buildReactor(reactorFactory=lambda: gireactor.GIReactor(useGtk=False))
```
